### ISSUE TITLE
Clang aborts in parsing the register pointer array (bug report)

### ISSUE THREAD

**Issue by drzchen** [llvm/llvm-project#56535]
When I use GCC to compile the following code, the compilation succeeds and the result is correct.
However, when I use Clang 14.0.6, a compilation error is reported and the compilation aborts.
I believe it is a bug of Clang (likely in the parsing module), as `rq[0]` is not the address of a register variable, but the address of array `a`.
```c
#include <stdio.h>
int main()
{
  int a[10] = {0,1,2,3,4,5};
  register int *rq[2] = { a }; /* rq[0] = a */
  printf("%d\n", *(rq[0]+1)); /* print 1 */
  return 0;
}
```
Clang reports the error:
```
testClang.c:6:20: error: address of register variable requested
  printf("%d\n", *(rq[0]+1)); /* print 1 */
                   ^~
1 error generated.
```

**Comment by phyBrackets** [comment]
Hey, the code works fine for me, the error you mentioned is in the case when you write `*(&rq[0]+1)` and which is pretty much correct you can't access the address for register variable in C .

**Comment by drzchen** [comment]
But what I write is `*(rq[0]+1)`, not your `*(&rq[0]+1)` (I know `&rq[0]` is not allowed).
Note that `*(rq[0]+1)` first gets the value at `rq[0]`, which equals `a` (note that `a` is an array in the stack, not register).
This means, `*(rq[0]+1)` is actually `*(a+1)`, which equals 1.
To conclude, `*(rq[0]+1)` never accesses the address of a register variable, and thus the error reported by clang 14.0.6 is a false positive.

**Comment by phyBrackets** [comment]
Ah I see I was on the previous version of the clang , so it was working for me . But `*(rq[0]+1)` also have problem. generally compilers converts the array index with pointer ,let's see how compiler sees the `rq[0]` it will convert it to `*(rq+0)` and it means you are trying to look into the address for `rq` and derefrencing it .

**Comment by phyBrackets** [comment]
A note from C standard https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2432.pdf , 

`128)The implementation can treat any register declaration simply as an auto declaration. However, whether or not addressable storage is actually used, the address of any part of an object declared with storage-class specifier register cannot be computed, either explicitly (by use of the unary & operator as discussed in 6.5.3.2) or implicitly (by converting an array name to a pointer as discussed in 6.3.2.1). Thus, the only operator that can be applied to an array declared with storage-class specifier register is sizeof.` 

And also the 'register' storage class specifier is deprecated and incompatible with C++17 .
